From de9c16d9c0bc85604570070cf07b2ad7c3a8b956 Mon Sep 17 00:00:00 2001
From: Mahaboob Vali <mvali@intrinsyc.com>
Date: Mon, 20 Aug 2018 10:29:23 +0530
Subject: [PATCH 007/334] [6857] input: pm8941-pwrkey: added sleep key event

This patch sends KEY_SLEEP event to userspace on short press of
power key. We are starting the timer when the key is pressed and
differentiated between long key press and short key press using
timer. If the key is released after 2 seconds, then it is
condsidered as long press and KEY_POWER is sent to the user space.
If the key is released before 2 seconds, then it is considered as
short key press, the timer is deleted  and KEY_SLEEP is sent to
the user space.

Change-Id: Id84d2bba11daa46b5d2df3653623925d082c7865
---
 drivers/input/misc/pm8941-pwrkey.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/input/misc/pm8941-pwrkey.c b/drivers/input/misc/pm8941-pwrkey.c
index 6671863bf65a..05f173d7cf97 100644
--- a/drivers/input/misc/pm8941-pwrkey.c
+++ b/drivers/input/misc/pm8941-pwrkey.c
@@ -149,6 +149,10 @@ static irqreturn_t pm8941_pwrkey_irq(int irq, void *_data)
 		mod_timer(&pwrkey->shutdown_timer, jiffies + msecs_to_jiffies(2000));
 	} else {
 		del_timer(&pwrkey->shutdown_timer);
+		input_event(pwrkey->input, EV_KEY, KEY_SLEEP, 1);
+		input_sync(pwrkey->input);
+		input_event(pwrkey->input, EV_KEY, KEY_SLEEP, 0);
+		input_sync(pwrkey->input);
 	}
 
 	return IRQ_HANDLED;
@@ -252,6 +256,7 @@ static int pm8941_pwrkey_probe(struct platform_device *pdev)
 	}
 
 	input_set_capability(pwrkey->input, EV_KEY, pwrkey->code);
+	input_set_capability(pwrkey->input, EV_KEY, KEY_SLEEP);
 
 	pwrkey->input->name = "pm8941_pwrkey";
 	pwrkey->input->phys = "pm8941_pwrkey/input0";
-- 
2.24.1

